src/prefect/agent.py

```diff
@@ -194,25 +194,67 @@ async def get_infrastructure(self, flow_run: FlowRun) -> Infrastructure:
 
     async def submit_run(self, flow_run: FlowRun) -> None:
         """
-        Submit a flow run to the flow runner
+        Submit a flow run to the infrastructure
         """
         ready_to_submit = await self._propose_pending_state(flow_run)
 
         if ready_to_submit:
             infrastructure = await self.get_infrastructure(flow_run)
 
-            try:
-                # Wait for submission to be completed. Note that the submission function
-                # may continue to run in the background after this exits.
-                await self.task_group.start(infrastructure.run)
-                self.logger.info(f"Completed submission of flow run '{flow_run.id}'")
-            except Exception as exc:
+            # Wait for submission to be completed. Note that the submission function
+            # may continue to run in the background after this exits.
+            await self.task_group.start(
+                self._submit_run_and_capture_errors, flow_run, infrastructure
+            )
+            self.logger.info(f"Completed submission of flow run '{flow_run.id}'")
+
+        self.submitting_flow_run_ids.remove(flow_run.id)
+
+    async def _submit_run_and_capture_errors(
+        self,
+        flow_run: FlowRun,
+        infrastructure: Infrastructure,
+        task_status: TaskStatus = None,
+    ) -> Union[InfrastructureResult, Exception]:
+
+        # Note: There is not a clear way to determine if task_status.started() has been
+        #       called without peeking at the internal `_future`. Ideally we could just
```

Comment:
Maybe throw a PR at `anyio` that exposes the `canceled`, `running`, and `done` methods on the Future or the future itself?

Reply:
I think this pattern of passing `task_status` through several functions would be considered depraved by them :D I'm hoping that this is resolved by splitting `run` into two separate methods: `submit` and `watch`.